File: 20. 有效的括号.js
/**
 * @param {string} s
 * @return {boolean}
 */

var isValid = function(s) {
    const m = {
        ')':'(',
        ']':'[',
        '}':'{'
    }
    var stack = [];
    for (var i = 0; i < s.length; i++) {
        var now = s[i];
        if (now == '(' || now == '[' || now == '{') stack.push(now);
        else if (now == ')' || now == ']' || now == '}') {
            if (stack.length <= 0) return false;
            var top = stack.pop();
            if (top != m[now]) return false;
        }
    }
    if (stack.length != 0) return false;
    else return true;
};

var s = "{[]}"
console.log(isValid(s))

File: 20. 有效的括号.cpp
class Solution {
public:
    bool isLeft(char c) {
        return c == '(' || c == '[' || c == '{';
    }
    bool isValid(string s) {
        map<char, char> m = {
            {'(', ')'},
            {'[', ']'},
            {'{', '}'}
        };
        stack<char> bracket_stack;

        for (int i = 0; i < s.length(); i++) {
            if (isLeft(s[i])) {
                bracket_stack.push(s[i]);
            } else {
                if (bracket_stack.size() == 0) return false;
                char left = bracket_stack.top();
                bracket_stack.pop();
                if (m[left] == s[i]) {
                    continue;
                } else {
                    return false;
                }
            }
        }

        return !bracket_stack.size();
    }
};